### PR TITLE
Returning session's state

### LINKED ISSUE #4784: Add SyncSession.getState()
Swift exposes `SyncSession.getState()` that exposes information about the sessions lifecycle. We should do something similar.

### PR DESCRIPTION
**Pull request by nhachicha**
fixes #4784

### CODE REVIEW

**Comment by nhachicha** [comment]
@realm/java

File: realm/realm-library/src/objectServer/java/io/realm/SyncSession.java [hunk 1]
```diff
@@ -86,6 +86,37 @@
     // we register the listener.
     private final AtomicLong progressListenerId = new AtomicLong(-1);
 
+    private static final byte STATE_VALUE_WAITING_FOR_ACCESS_TOKEN = 0;
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/SyncSession.java]: Perhaps add a reference to the C++ code where these are defined?

File: realm/realm-library/src/objectServer/java/io/realm/SyncSession.java [hunk 2]
```diff
@@ -86,6 +86,37 @@
...
+    private static final byte STATE_VALUE_ACTIVE = 1;
+    private static final byte STATE_VALUE_DYING = 2;
+    private static final byte STATE_VALUE_INACTIVE = 3;
+    private static final byte STATE_VALUE_ERROR = 4;
+
+    public enum State {
+        STATE_WAITING_FOR_ACCESS_TOKEN(STATE_VALUE_WAITING_FOR_ACCESS_TOKEN),
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/SyncSession.java]: No reason to prefix these with `STATE_`. The enum is already called `State`.

File: realm/realm-library/src/objectServer/java/io/realm/SyncSession.java [hunk 3]
```diff
@@ -86,6 +86,37 @@
...
+        STATE_INACTIVE(STATE_VALUE_INACTIVE),
+        STATE_ERROR(STATE_VALUE_ERROR);
+
+        final byte value;
+
+        State(byte value) {
+            this.value = value;
+        }
+
+        public static State fromByte(byte value) {
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/SyncSession.java]: This method is in the public API. Is that really necessary? If yes, it need Javadoc

File: realm/realm-library/src/objectServer/java/io/realm/SyncSession.java [hunk 4]
```diff
@@ -136,6 +167,17 @@ void notifySessionError(int errorCode, String errorMessage) {
         }
     }
 
+    @KeepMember
+    @SuppressWarnings("unused")
+    public State getState() {
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/SyncSession.java]: Missing Javadoc

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/SyncSessionTests.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    public void getState_active() {
+        SyncUser user = UserFactory.createUniqueUser(Constants.AUTH_URL);
+        SyncConfiguration syncConfiguration = configFactory
+                .createSyncConfigurationBuilder(user, Constants.SYNC_SERVER_URL)
+                .waitForInitialRemoteData()
+                .build();
+        Realm realm = Realm.getInstance(syncConfiguration);
+
+        SyncSession session = SyncManager.getSession(syncConfiguration);
+        assertEquals(SyncSession.State.STATE_ACTIVE, session.getState());
```
Comment (cmelchior) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/SyncSessionTests.java]: Isn't there a timing issue involved here, e.g if the server is slow returning the access token?
I'm not sure how to solve this though

File: realm/realm-library/src/objectServer/java/io/realm/SyncSession.java (hunk 3 above)
Comment (nhachicha) [inline realm/realm-library/src/objectServer/java/io/realm/SyncSession.java]: 👍 package private now

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/SyncSessionTests.java (hunk 5 above)
Comment (nhachicha) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/SyncSessionTests.java]: you're right we could still be waiting for the `access_token` to return & be in `WAITING_FOR_ACCESS_TOKEN` state. Will add a small delay as an empirical solution

Comment (cmelchior) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/SyncSessionTests.java]: Perhaps also use https://developer.android.com/reference/android/support/test/filters/FlakyTest.html ?

File: CHANGELOG.md [hunk 6]
```diff
@@ -7,6 +7,7 @@
 ### Enhancements
 
 * [ObjectServer] Added support for Sync Progress Notifications through `SyncSession.addDownloadProgressListener(ProgressMode, ProgressListener)` and `SyncSession.addUploadProgressListener(ProgressMode, ProgressListener)` (#4104).
+* [ObjectServer] Added `Session.getState()` (#4784).
```
Comment (cmelchior) [inline CHANGELOG.md]: SyncSession

File: realm/realm-library/src/objectServer/java/io/realm/SyncSession.java [hunk 7]
```diff
@@ -589,6 +642,7 @@ public void throwExceptionIfNeeded() {
 
     private static native long nativeAddProgressListener(String localRealmPath, long listenerId, int direction, boolean isStreaming);
     private static native void nativeRemoveProgressListener(String localRealmPath, long listenerToken);
-    private static native boolean nativeRefreshAccessToken(String path, String accessToken, String realmUrl);
-    private native boolean nativeWaitForDownloadCompletion(String path);
+    private static native boolean nativeRefreshAccessToken(String localRealmPath, String accessToken, String realmUrl);
+    private native boolean nativeWaitForDownloadCompletion(String localRealmPath);
+    private native byte nativeGetState(String localRealmPath);
```
Comment (beeender) [inline realm/realm-library/src/objectServer/java/io/realm/SyncSession.java]: these two can be `static`.

**Pull request merged**